### PR TITLE
Changes to typing module rules

### PR DESCRIPTION
**Pull request by ngtvspc**
I've made some changes in this PR to the way that the py3 builtin module `typing` is handled. `typing` was [backported](https://pypi.org/project/typing/) to work on older versions of Python. The trove classifiers on PyPI list 2.7 and 3.4 as supported versions. I suspect it will import on as low as 3.1 based on `vermin`s output and my understanding of `typing`.

Currently this PR just adds 2.7 as a supported version for the `typing` module and the attributes this backport provides. If it seems acceptable I'd like to also change the support for the rules I modified to lower the py3 version to 3.1. 

Analyzing `typing` installed via `pip` using `vermin`
```
(vermin_env) shawn@deepthink:~/Documents/vermin$ vermin -vv ../vermin_env/lib/python2.7/site-packages/typing.py
Detecting python files..
Analyzing using 8 processes..
!2, 3.3      /home/shawn/Documents/vermin_env/lib/python2.7/site-packages/typing.py
  '__future__' requires 2.1, 3.0
  '__future__.absolute_import' requires 2.5, 3.0
  '__future__.unicode_literals' requires 2.6, 3.0
  'abc' requires 2.6, 3.0
  'all' requires 2.5, 3.0
  'any' requires 2.5, 3.0
  'bool' requires 2.2, 3.0
  'classmethod' requires 2.2, 3.0
  'collections' requires 2.4, 3.0
  'collections.Counter' requires 2.7, 3.1
  'collections.abc' requires !2, 3.3
  'collections.defaultdict' requires 2.5, 3.0
  'collections.deque' requires 2.4, 3.0
  'collections.namedtuple' requires 2.6, 3.0
  'enumerate' requires 2.3, 3.0
  'frozenset' requires 2.4, 3.0
  'functools' requires 2.5, 3.0
  'set' requires 2.4, 3.0
  'super' requires 2.2, 3.0
  'sys.version_info' requires 2.0, 3.0
  'type' requires 2.2, 3.0
  'zip' requires 2.0, 3.0
  `"..{}..".format(..)` requires (2.7, 3.0)
  dict comprehensions require (2.7, 3.0)

Minimum required versions: 3.3
Incompatible versions:     2
```

The only py2 incompatible line is `'collections.abc' requires !2, 3.3` and this is the associated code in `typing`:

```
from __future__ import absolute_import, unicode_literals

import abc
from abc import abstractmethod, abstractproperty
import collections
import functools
import re as stdlib_re  # Avoid confusion with the re we export.
import sys
import types
import copy
try:
    import collections.abc as collections_abc
except ImportError:
    import collections as collections_abc  # Fallback for PY3.2.


# Please keep __all__ alphabetized within each category.
__all__ = [
    # Super-special typing primitives.
    'Any',
...
```

~As a minor thing I also fixed a test that failed on my machine on the master branch. I think it's just a simple ordering issue of the list that comes back~ (Looks like my changes using `assertItemsEqual` aren't compatible with the versions of python CI runs against, so I momentarily reverted them)

### CODE REVIEW

**Comment by netromdk** [comment]
Thank you for the PR!

My rules detect standard Python modules and features. In Python 2.7.17, I get `ImportError` on `import typing`, which underscores why `(2, 7)` isn't in the rules. If one wants to install a PyPi package for support then that's fine. It is an interesting question on whether to include PyPi info but since it's not official Python code, I'm not sure it feels right to do so.

Also, the list of modules from the 2.7 docs: https://docs.python.org/2.7/py-modindex.html#cap-t
And from the [3.x docs](https://docs.python.org/3/library/typing.html?highlight=typing#module-typing):
> New in version 3.5.

In a case such as this, if you install a backported package, it might make sense to use [`--exclude typing`](https://github.com/netromdk/vermin#analysis-exclusions) when running Vermin since it's non-standard Python all of a sudden.

**Comment by ngtvspc** [comment]
That's fair, it comes down to how you want to handle it. The backported version is definitely not in the standard library for 2.7 and I think drawing the line at only modules and features available in the standard library is reasonable.

The particular issue here is that a package from PyPI is causing false positives among the rules used in `vermin`. There might be other ways to address this, instead of changing the rules how would you feel about `vermin` parsing an optional `requirements.txt`?

**Comment by netromdk** [comment]
What would you put in the file and when would it be used?

Btw, I don't get the issue you had with the test of files. The sorting yields a uniform ordering so if it fails it might be the elements in the lists that are different instead. Could you show me the test output when it fails for you? (only for the test case and the text related). Thanks.

**Comment by ngtvspc** [comment]
```

File: tests/general.py [hunk 1]
```diff
@@ -97,12 +97,12 @@ def test_detect_hidden_paths(self):
              touch(tmp_fld, "test2.py")]
 
     paths = detect_paths([tmp_fld], hidden=False)
-    self.assertEqual([files[1], files[3]], paths)
+    self.assertItemsEqual([files[1], files[3]], paths)
```
Comment (netromdk) [inline tests/general.py]: Ah, the fix here is to do:
```py
paths.sort()
specific_files = [files[1], files[3]]
specific_files.sort()
```
after line 99 and before line 100.

Comment (netromdk) [reply tests/general.py]: And then keep `self.assertEqual(..)`

Comment (ngtvspc) [reply tests/general.py]: Yeah `assertItemsEqual` just isn't part of the API in later versions of Pythons unittest. I just wanted to make sure I hadn't broken anything else with my changes to the rules

Comment (netromdk) [reply tests/general.py]: Another solution would be to add `assertItemsEqual()` to `VerminTest` (in "tests/testutils.py")? That would actually be nice.

Comment (netromdk) [reply tests/general.py]: Since the sorting, in both cases, wouldn't be necessary.

**Comment by ngtvspc** [comment]
As for the parsing of a `requirements.txt`. I find myself often using `vermin` to analyze full applications and the output can often be a little confusing because it will contradict what I know about the application. It usually comes down to a couple libraries like `typing` that have backports and trigger the rules.

I think many applications manage their dependencies using a `requirements.txt`, and it's usually in the projects root directory so it wouldn't be too hard to specify.

I see two proposals:
- auto detect the `requirements.txt` and warn the user if it contains libraries that may trigger rules
- only use a `requirements.txt` if the user manually specifies it and use something similar to `--exclude` on conflicting libraries

I've got to sign off for the night, it's pretty late in California. Thanks for taking the time to discuss some ideas with me

**Comment by netromdk** [comment]
I like the idea of auto-detecting a "requirements.txt" (and maybe adding a `--no-req-file` to disable it), and an explicit file to use via `--req-file <path>`.

The list of potentially problematic modules could be added to "vermin/constants.py" as `NON_STD_REQ_MODULES = ["typing"]` perhaps.

Does it then make most sense to exclude "typing", for example, in rules or to still apply the rules. But show the warning in any case?

Thanks for taking time to discuss! :)

**Comment by ngtvspc** [comment]
Thanks for the tip about using `--exclude` by the way, I think I'll incorporate that into how we're tracking py2 and py3 files at work and that should work for now.

I think if `vermin` auto-detects a requirements file it should probably still apply all the rules. Mostly because I would like to prevent any change I make from altering how users expect it to behave in significant ways.

But if a user explicitly provides a requirements file I think it would be okay to assume they want to suppress any rules that could be interpreted as ambiguous like `typing` in my case. I'll discuss some of these ideas with a few people I know at work who have a lot of experience writing and maintaining command line applications just to make sure I'm not missing something obvious. After I can find a little time to push out a prototype of it I think exactly how it should behave will become easier to pin down

**Comment by netromdk** [comment]
You're welcome!

You want to look into the requirements file feature? That's great. Looking forward to that. Will you push to this PR or another one?

In the meantime, I'm going to either sort the values we talked about or implement the `assertItemsEqual()` and use that.

**Comment by netromdk** [comment]
As an aside, `assertItemsEqual()` is a 2.7 `unittest` function. In 3.2 it's called `assertCountEqual()`. That's probably why it failed for you. Making one called `assertItemsEqual()` that works with 2.7, 3.2+, and 3.0-3.1.

**Comment by ngtvspc** [comment]
I'll look into adding the requirements feature, I'll push it to another PR. I'll close this one for now

**Comment by ngtvspc** [comment]
Also the change you made to the tests now pass on my machine, thanks. ^.^

**Comment by netromdk** [comment]
Thanks, looking forward to it. :)

**Pull request closed without merging**